crates/ruma-identifiers/src/client_secret.rs

```diff
@@ -1,164 +1,16 @@
 //! Client secret identifier.
 
-use std::{convert::TryFrom, fmt, mem, rc::Rc, str::FromStr, sync::Arc};
-
 use ruma_identifiers_validation::client_secret::validate;
 
-/// A client secret.
-///
-/// Session IDs in Matrix are opaque character sequences of `[0-9a-zA-Z.=_-]. Their length must
-/// must not exceed 255 characters.
-#[repr(transparent)]
-#[derive(PartialEq, Eq, PartialOrd, Ord, Hash)]
-#[cfg_attr(feature = "serde", derive(serde::Serialize), serde(transparent, crate = "serde"))]
-pub struct ClientSecret(str);
-
-/// An owned client secret.
-pub type ClientSecretBox = Box<ClientSecret>;
-
-impl ClientSecret {
-    #[allow(clippy::transmute_ptr_to_ptr)]
-    fn from_borrowed(s: &str) -> &Self {
-        unsafe { mem::transmute(s) }
-    }
-
-    fn from_owned(s: Box<str>) -> Box<Self> {
-        unsafe { mem::transmute(s) }
-    }
-
-    fn into_owned(self: Box<Self>) -> Box<str> {
-        unsafe { mem::transmute(self) }
-    }
-
-    /// Creates a string slice from this `ClientSecret`.
-    pub fn as_str(&self) -> &str {
-        &self.0
-    }
-
-    /// Creates a byte slice from this `ClientSecret`.
-    pub fn as_bytes(&self) -> &[u8] {
-        self.0.as_bytes()
-    }
-}
-
-impl fmt::Debug for ClientSecret {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_str(self.as_str())
-    }
-}
-
-impl Clone for Box<ClientSecret> {
-    fn clone(&self) -> Self {
-        (**self).to_owned()
-    }
-}
-
-impl ToOwned for ClientSecret {
-    type Owned = Box<ClientSecret>;
-
-    fn to_owned(&self) -> Self::Owned {
-        Self::from_owned(self.0.to_owned().into_boxed_str())
-    }
-}
-
-impl From<&ClientSecret> for Box<ClientSecret> {
-    fn from(s: &ClientSecret) -> Self {
-        s.to_owned()
-    }
-}
-
-impl From<&ClientSecret> for Rc<ClientSecret> {
-    fn from(s: &ClientSecret) -> Self {
-        let rc = Rc::<str>::from(s.as_str());
-        unsafe { Rc::from_raw(Rc::into_raw(rc) as *const ClientSecret) }
-    }
-}
-
-impl From<&ClientSecret> for Arc<ClientSecret> {
-    fn from(s: &ClientSecret) -> Self {
-        let arc = Arc::<str>::from(s.as_str());
-        unsafe { Arc::from_raw(Arc::into_raw(arc) as *const ClientSecret) }
-    }
-}
-
-fn try_from<S>(client_secret: S) -> Result<Box<ClientSecret>, crate::Error>
-where
-    S: AsRef<str> + Into<Box<str>>,
-{
-    validate(client_secret.as_ref())?;
-    Ok(ClientSecret::from_owned(client_secret.into()))
+opaque_identifier_validated! {
+    /// A client secret.
+    ///
+    /// Client secrets in Matrix are opaque character sequences of `[0-9a-zA-Z.=_-]`. Their length must
+    /// must not exceed 255 characters.
+    pub type ClientSecret;
+    validate;
```

Comment:
I don't like `validate` looking like a separate declaration, though I'm not sure what to suggest instead. Maybe

```suggestion
    pub type ClientSecret [ validate ];
```

cc @iinuwa